markdown/how.md

```diff
@@ -3,17 +3,87 @@
 
 ## OpenStack
 
+<!-- Note -->
+
+Our cloud infrastructure runs on OpenStack, and we use it for both the infrastructure of the learning platform and our lab environments.
+
+With the native installation, we used Openstack Heat to deploy our learning platforms to OpenStack VMs. After transitioning to Docker-based deployment, i.e., containerizing the services and running them on Kubernetes, we use Opnetsack magnum to deploy our Kubernetes cluster.
+
+Additionally, we run a self-paced, interactive learning platform, meaning we provide the learners with access to arbitrarily complex, realistic distributed environments on demand where they can learn things by actually doing it and OpenStack Heat comes in handy. 
+Using a Heat template, we can define an exactly reproducible, self-contained environment consisting of, say, ten servers in three networks connected with two routers and arbitrarily involved configurations for each server. 
+
+And heat also provides the ability to suspend the entire heat stack and then resume it at a much later date in the exact same state, which makes our platform very cost-effective.
+
 
 ## Magnum
 
+<!-- Note -->
+
+In order to transition to the Docker-based deployment model with Kubernetes, we required a container orchestration service and a container registry offering for storing private container images.
+Within our cloud infrastructure, we had two options available for container orchestration services: Rancher and Magnum. We chose OpenStack Magnum because it provided all the necessary components for a production-ready setup for deploying Tutor.
+
 
 ### Private registry
 
+<!-- Note -->
+
+We needed a container orchestration service that could accommodate custom-built Docker images, and thus we decided to utilize the private Docker registry provided by Magnum. The built-in Docker registry runs locally on each Kubernetes's worker nodes localhost. It does not require authentication, making it easy to publish and pull images without any configuration changes to Tutor.
+
+The Docker registry supported by Magnum is just the official Docker registry image with the OpenStack Swift storage driver. When invoked by Magnum, it is Magnum that automatically populates the registry's storage configuration with the necessary parameters to match the Swift endpoint in the region it's running in. In our CI, we duplicate this for a locally-running registry instance, backed by the same storage parameters, that we push our images to. Once we do, any thus-registered image becomes available to our Magnum registries.
+
+In our case it's not native OpenStack Swift that backs the registry, but Ceph radosgw with Keystone integration.
+
+We don't just rely on a single registry, but instead, we have two registries one for each region. All images are available in both regions, ensuring redundancy and availability across our infrastructure.
+
+
```

Comment:
Of these two blank lines, drop one.

Reply:
Dropping this blank line messes with the layout. I see notes inside the slides.